File: server/bot_ai.py
"""
2048 — Sunucu tarafı BOT motoru (expectimax).

İstemcideki src/app/logic/ai.ts içindeki bot fonksiyonlarının (botMove /
playBotGame) BİREBİR Python eşi. Çok oyunculu yarışta bot artık SUNUCUDA koşar:
tohumdan botun tüm oyununu bir kez oynatıp skor ZAMAN ÇİZELGESİ üretiriz; yarış
boyunca o çizelge yayınlanır (gerçek zamanlı hesap yok, host'tan/ cihazdan
bağımsız, manipüle edilemez).

DETERMİNİZM & PARİTE:
  - Sabit derinlik (zaman sınırı yok) → makineden bağımsız.
  - Ağırlıklar TAM SAYIYA yuvarlanır → tek transandantal (pow) işlemi sabitlenir;
    geri kalan her şey tam-sayı/IEEE aritmetiğidir → JS ile birebir aynı sonuç.
  Parite server/test_bot_parity.py (TS'in ürettiği bot_fixtures.json) ile korunur.
"""
import math

from replay import mulberry32, move_grid, _empty_cells, _spawn, _max, _MASK

CHANCE_OF_FOUR = 0.1
NO_MOVE_PENALTY = -1e15
DIRECTIONS = ("up", "down", "left", "right")
DIR_CHAR = {"up": "U", "down": "D", "left": "L", "right": "R"}

# Botun hamle temposu (ms) — güçlü seviye daha hızlı/akıcı oynar. Sunucu skor
# çizelgesini bu tempoya göre yayınlar (eski istemci speedFor'un sunucu karşılığı).
# Karakterler de kendi temposunu taşır ("Acelesi Var" en hızlı oynar).
BOT_SPEED_MS = {
    "easy": 480, "medium": 360, "hard": 280, "expert": 240,
    "corner": 300, "space": 330, "hasty": 240, "balanced": 300,
}

# Seviye başına arama ayarları — ai.ts LEVEL_CFG + BOT_DEPTH ile BİREBİR aynı.
BOT_CFG = {
    "easy":   {"depth": 1, "sampleK": 2, "expandFour": True,  "snakePow": 1.0, "emptyMul": 4},
    "medium": {"depth": 2, "sampleK": 1, "expandFour": False, "snakePow": 0.3, "emptyMul": 1},
    "hard":   {"depth": 2, "sampleK": 2, "expandFour": True,  "snakePow": 1.0, "emptyMul": 4},
    "expert": {"depth": 3, "sampleK": 2, "expandFour": True,  "snakePow": 1.0, "emptyMul": 4},
}

# Bot KARAKTERLERİ — ai.ts BOT_CHARACTERS ile BİREBİR aynı ağırlık setleri.
# Motor değişmez; her karakter yalnızca farklı bir ağırlık seti (snakePow/emptyMul
# + derinlik/örnekleme). Parite: server/test_bot_parity.py (karakter fixture'ları).
BOT_CHARACTERS = {
    "corner":   {"depth": 2, "sampleK": 2, "expandFour": True, "snakePow": 1.2, "emptyMul": 1},
    "space":    {"depth": 2, "sampleK": 2, "expandFour": True, "snakePow": 0.5, "emptyMul": 8},
    "hasty":    {"depth": 1, "sampleK": 2, "expandFour": True, "snakePow": 0.3, "emptyMul": 1},
    "balanced": {"depth": 2, "sampleK": 2, "expandFour": True, "snakePow": 1.0, "emptyMul": 4},
}


def resolve_cfg(key):
    """Bir bot anahtarını (zorluk kademesi VEYA karakter kimliği) ayara çözer.
    ai.ts resolveBotConfig eşi. Sunucu `level` sütununda ikisini de tutabilir."""
    if key in BOT_CHARACTERS:
        return BOT_CHARACTERS[key]
    return BOT_CFG.get(key, BOT_CFG["medium"])

_weight_cache = {}


def snake_weights(n, pow):
    """ai.ts snakeWeights ile birebir: yılan gradyanı, floor(base^(idx·pow)+0.5)."""
    key = (n, pow)
    cached = _weight_cache.get(key)
    if cached is not None:
        return cached
    base = 3 if n >= 5 else 4
    w = [[0] * n for _ in range(n)]
    idx = n * n - 1
    for r in range(n):
        cols = range(n) if r % 2 == 0 else range(n - 1, -1, -1)
        for c in cols:
            w[r][c] = int(math.floor(math.pow(base, idx * pow) + 0.5))
            idx -= 1
    _weight_cache[key] = w
    return w


def evaluate(g, snake_pow, empty_mul):
    """Izgarayı puanlar — ai.ts evaluate ile birebir (tam-sayı)."""
    n = len(g)
    w = snake_weights(n, snake_pow)
    weighted = 0
    empties = 0
    max_val = 0
    for r in range(n):
        for c in range(n):
            v = g[r][c]
            if v == 0:
                empties += 1
                continue
            weighted += v * w[r][c]
            if v > max_val:
                max_val = v
    return weighted + empties * max_val * empty_mul


def _sample_cells(cells, k):
    """ai.ts sampleCells ile birebir: eşit aralıklı k hücre (deterministik)."""
    step = len(cells) / k
    out = []
    for i in range(k):
        out.append(cells[int(math.floor(i * step))])
    return out


def _place(g, r, c, v):
    out = [row[:] for row in g]
    out[r][c] = v
    return out


def _max_node(g, depth, cfg):
    if depth == 0:
        return evaluate(g, cfg["snakePow"], cfg["emptyMul"])
    best = None
    for d in DIRECTIONS:
        nxt, _gained, moved = move_grid(g, d)
        if not moved:
            continue
        v = _chance_node(nxt, depth - 1, cfg)
        if best is None or v > best:
            best = v
    return NO_MOVE_PENALTY if best is None else best


def _chance_node(g, depth, cfg):
    if depth == 0:
        return evaluate(g, cfg["snakePow"], cfg["emptyMul"])
    cells = _empty_cells(g)
    if not cells:
        return evaluate(g, cfg["snakePow"], cfg["emptyMul"])
    k = cfg["sampleK"]
    sample = _sample_cells(cells, k) if len(cells) > k else cells
    total = 0.0
    per = 1 / len(sample)
    for (r, c) in sample:
        if cfg["expandFour"]:
            total += per * 0.9 * _max_node(_place(g, r, c, 2), depth, cfg)
            total += per * CHANCE_OF_FOUR * _max_node(_place(g, r, c, 4), depth, cfg)
        else:
            total += per * _max_node(_place(g, r, c, 2), depth, cfg)
    return total


def bot_move_cfg(g, cfg):
    """Sabit derinlikte deterministik expectimax hamlesi. ai.ts botMoveCfg eşi."""
    legal = [d for d in DIRECTIONS if move_grid(g, d)[2]]
    if not legal:
        return None
    if len(legal) == 1:
        return legal[0]
    best = legal[0]
    best_val = None
    for d in legal:
        nxt, _gained, _moved = move_grid(g, d)
        v = _chance_node(nxt, cfg["depth"] - 1, cfg)
        if best_val is None or v > best_val:
            best_val = v
            best = d
    return best


def bot_move(g, level):
    """Zorluk kademesi/karakteriyle bot hamlesi (geriye dönük). ai.ts botMove eşi."""
    return bot_move_cfg(g, resolve_cfg(level))


def play_bot_game(seed, level, max_moves, size=4, on_progress=None, progress_every=32):
    """
    Tohumdan botun oyununu (en çok max_moves hamle) oynatır; skor ZAMAN
    çizelgesini döndürür. ai.ts playBotGame/playBotGameByKey ile birebir.
    `level` bir zorluk kademesi VEYA karakter kimliği olabilir (resolve_cfg).
      scores[k] = k. hamleden SONRA kümülatif skor (scores[0] = 0)
      bests[k]  = k. hamleden sonra en büyük kare

    on_progress(scores, bests) verilirse her `progress_every` hamlede bir (ve
    oyun bitince) çağrılır → sunucu çizelgeyi ARTIMLI yayınlar (uzun süren Uzman
    hesabında bot yarış saatini geçtiği için hiç aç kalmaz). Callback OYUNU
    ETKİLEMEZ; çıktı parite testinde birebir aynıdır.
    """
    cfg = resolve_cfg(level)
    rand = mulberry32(seed & _MASK)
    grid = [[0] * size for _ in range(size)]
    _spawn(grid, rand)
    _spawn(grid, rand)

    moves = []
    scores = [0]
    bests = [_max(grid)]
    score = 0
    for i in range(max_moves):
        d = bot_move_cfg(grid, cfg)
        if d is None:
            break
        nxt, gained, moved = move_grid(grid, d)
        if not moved:
            break
        grid = nxt
        score += gained
        moves.append(DIR_CHAR[d])
        _spawn(grid, rand)
        scores.append(score)
        bests.append(_max(grid))
        if on_progress is not None and (i + 1) % progress_every == 0:
            on_progress(scores, bests)

    if on_progress is not None:
        on_progress(scores, bests)
    return {
        "moves": "".join(moves),
        "scores": scores,
        "bests": bests,
        "maxTile": _max(grid),
        "finalScore": score,
    }

File: api-nest/src/rooms/bot-ai.ts
// ============================================================
//  2048 — Sunucu tarafı BOT motoru (expectimax) — bot_ai.py BİREBİR portu.
//  İstemci ai.ts botMove/playBotGame'in eşi. Sabit derinlik + tam-sayıya
//  yuvarlanmış ağırlıklar → tek transandantal (pow) sabitlenir, gerisi JS ile
//  birebir. Parite: test/bot.parity.spec.ts (server/bot_fixtures.json).
// ============================================================

import { mulberry32, moveGrid, emptyCells, spawn, maxOf, Direction } from '../replay/replay';

const CHANCE_OF_FOUR = 0.1;
const NO_MOVE_PENALTY = -1e15;
const DIRECTIONS: Direction[] = ['up', 'down', 'left', 'right'];
const DIR_CHAR: Record<Direction, string> = { up: 'U', down: 'D', left: 'L', right: 'R' };

export interface BotCfg {
  depth: number;
  sampleK: number;
  expandFour: boolean;
  snakePow: number;
  emptyMul: number;
}

/** Hamle temposu (ms) — güçlü seviye daha hızlı. app.py BOT_SPEED_MS birebir. */
export const BOT_SPEED_MS: Record<string, number> = {
  easy: 480, medium: 360, hard: 280, expert: 240,
  corner: 300, space: 330, hasty: 240, balanced: 300,
};

const BOT_CFG: Record<string, BotCfg> = {
  easy: { depth: 1, sampleK: 2, expandFour: true, snakePow: 1.0, emptyMul: 4 },
  medium: { depth: 2, sampleK: 1, expandFour: false, snakePow: 0.3, emptyMul: 1 },
  hard: { depth: 2, sampleK: 2, expandFour: true, snakePow: 1.0, emptyMul: 4 },
  expert: { depth: 3, sampleK: 2, expandFour: true, snakePow: 1.0, emptyMul: 4 },
};

/** Bot karakterleri — ai.ts BOT_CHARACTERS birebir. */
export const BOT_CHARACTERS: Record<string, BotCfg> = {
  corner: { depth: 2, sampleK: 2, expandFour: true, snakePow: 1.2, emptyMul: 1 },
  space: { depth: 2, sampleK: 2, expandFour: true, snakePow: 0.5, emptyMul: 8 },
  hasty: { depth: 1, sampleK: 2, expandFour: true, snakePow: 0.3, emptyMul: 1 },
  balanced: { depth: 2, sampleK: 2, expandFour: true, snakePow: 1.0, emptyMul: 4 },
};

/** Bot anahtarı bir karakter mi (level sütununda karakter/zorluk ayrımı). */
export function isCharacter(key: string | null | undefined): boolean {
  return !!key && key in BOT_CHARACTERS;
}

/** Anahtarı (zorluk VEYA karakter) ayara çözer. app.py resolve_cfg birebir. */
export function resolveCfg(key: string | null | undefined): BotCfg {
  if (key && key in BOT_CHARACTERS) return BOT_CHARACTERS[key];
  return BOT_CFG[key as string] ?? BOT_CFG['medium'];
}

const weightCache = new Map<string, number[][]>();

/** ai.ts snakeWeights: yılan gradyanı floor(base^(idx·pow)+0.5). */
function snakeWeights(n: number, pow: number): number[][] {
  const key = `${n}|${pow}`;
  const cached = weightCache.get(key);
  if (cached) return cached;
  const base = n >= 5 ? 3 : 4;
  const w = Array.from({ length: n }, () => new Array(n).fill(0));
  let idx = n * n - 1;
  for (let r = 0; r < n; r++) {
    const cols =
      r % 2 === 0
        ? Array.from({ length: n }, (_, i) => i)
        : Array.from({ length: n }, (_, i) => n - 1 - i);
    for (const c of cols) {
      w[r][c] = Math.floor(Math.pow(base, idx * pow) + 0.5);
      idx--;
    }
  }
  weightCache.set(key, w);
  return w;
}

/** Izgarayı puanlar — ai.ts evaluate birebir (tam-sayı). */
function evaluate(g: number[][], snakePow: number, emptyMul: number): number {
  const n = g.length;
  const w = snakeWeights(n, snakePow);
  let weighted = 0;
  let empties = 0;
  let maxVal = 0;
  for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
      const v = g[r][c];
      if (v === 0) {
        empties++;
        continue;
      }
      weighted += v * w[r][c];
      if (v > maxVal) maxVal = v;
    }
  }
  return weighted + empties * maxVal * emptyMul;
}

/** ai.ts sampleCells: eşit aralıklı k hücre (deterministik). */
function sampleCells(cells: Array<[number, number]>, k: number): Array<[number, number]> {
  const step = cells.length / k;
  const out: Array<[number, number]> = [];
  for (let i = 0; i < k; i++) out.push(cells[Math.floor(i * step)]);
  return out;
}

function place(g: number[][], r: number, c: number, v: number): number[][] {
  const out = g.map((row) => row.slice());
  out[r][c] = v;
  return out;
}

function maxNode(g: number[][], depth: number, cfg: BotCfg): number {
  if (depth === 0) return evaluate(g, cfg.snakePow, cfg.emptyMul);
  let best: number | null = null;
  for (const d of DIRECTIONS) {
    const { moved, grid: nxt } = moveGrid(g, d);
    if (!moved) continue;
    const v = chanceNode(nxt, depth - 1, cfg);
    if (best === null || v > best) best = v;
  }
  return best === null ? NO_MOVE_PENALTY : best;
}

function chanceNode(g: number[][], depth: number, cfg: BotCfg): number {
  if (depth === 0) return evaluate(g, cfg.snakePow, cfg.emptyMul);
  const cells = emptyCells(g);
  if (cells.length === 0) return evaluate(g, cfg.snakePow, cfg.emptyMul);
  const k = cfg.sampleK;
  const sample = cells.length > k ? sampleCells(cells, k) : cells;
  let total = 0;
  const per = 1 / sample.length;
  for (const [r, c] of sample) {
    if (cfg.expandFour) {
      total += per * 0.9 * maxNode(place(g, r, c, 2), depth, cfg);
      total += per * CHANCE_OF_FOUR * maxNode(place(g, r, c, 4), depth, cfg);
    } else {
      total += per * maxNode(place(g, r, c, 2), depth, cfg);
    }
  }
  return total;
}

/** Sabit derinlikte deterministik expectimax hamlesi. ai.ts botMoveCfg eşi. */
export function botMoveCfg(g: number[][], cfg: BotCfg): Direction | null {
  const legal = DIRECTIONS.filter((d) => moveGrid(g, d).moved);
  if (legal.length === 0) return null;
  if (legal.length === 1) return legal[0];
  let best = legal[0];
  let bestVal: number | null = null;
  for (const d of legal) {
    const { grid: nxt } = moveGrid(g, d);
    const v = chanceNode(nxt, cfg.depth - 1, cfg);
    if (bestVal === null || v > bestVal) {
      bestVal = v;
      best = d;
    }
  }
  return best;
}

export interface BotGameResult {
  moves: string;
  scores: number[];
  bests: number[];
  maxTile: number;
  finalScore: number;
}

/**
 * Tohumdan botun oyununu oynatır; skor ZAMAN çizelgesini döndürür.
 * app.py play_bot_game / ai.ts playBotGame birebir. `onProgress` verilirse
 * her `progressEvery` hamlede (ve bitişte) çağrılır (artımlı yayın).
 */
export function playBotGame(
  seed: number,
  level: string,
  maxMoves: number,
  size = 4,
  onProgress?: (scores: number[], bests: number[]) => void,
  progressEvery = 32,
): BotGameResult {
  const cfg = resolveCfg(level);
  const rand = mulberry32(seed >>> 0);
  let grid = Array.from({ length: size }, () => new Array(size).fill(0));
  spawn(grid, rand);
  spawn(grid, rand);

  const moves: string[] = [];
  const scores = [0];
  const bests = [maxOf(grid)];
  let score = 0;
  for (let i = 0; i < maxMoves; i++) {
    const d = botMoveCfg(grid, cfg);
    if (d === null) break;
    const res = moveGrid(grid, d);
    if (!res.moved) break;
    grid = res.grid;
    score += res.gained;
    moves.push(DIR_CHAR[d]);
    spawn(grid, rand);
    scores.push(score);
    bests.push(maxOf(grid));
    if (onProgress && (i + 1) % progressEvery === 0) onProgress(scores, bests);
  }
  if (onProgress) onProgress(scores, bests);
  return { moves: moves.join(''), scores, bests, maxTile: maxOf(grid), finalScore: score };
}

/** Yarış süresini kaplayacak hamle sayısı (+tampon). app.py _bot_max_moves. */
export function botMaxMoves(level: string, duration: number): number {
  const speed = BOT_SPEED_MS[level] ?? 360;
  return Math.floor((duration * 1000) / speed) + 6;
}
